armi/reactor/components/component.py

```diff
@@ -1262,6 +1264,72 @@ def getIntegratedMgFlux(self, adjoint=False, gamma=False):
 
         return pinFluxes[self.p.pinNum - 1] * self.getVolume()
 
+    def getPinMgFluxes(
+        self, adjoint: Optional[bool] = False, gamma: Optional[bool] = False
+    ) -> np.ndarray:
+        """Retrieves the pin multigroup fluxes for the component.
+
+        Parameters
+        ----------
+        adjoint : bool, optional
+            Return adjoint flux instead of real
+        gamma : bool, optional
+            Whether to return the neutron flux or the gamma flux.
+
+        Returns
+        -------
+        np.ndarray
+            A ``(N, nGroup)`` array of pin multigroup fluxes, where ``N`` is the
+            equivalent to the multiplicity of the component (``self.p.mult``)
+            and ``nGroup`` is the number of energy groups of the flux.
+
+        Raises
+        ------
+        ValueError
+            If the location(s) of the component are not aligned with pin indices
+            from the block. This would happen if this component is not actually
+            a pin.
+        """
+        # Get the (i, j, k) location of all pins from the parent block
+        # FIXME: This should be changed to just using Block.getPinLocations once Drew's PR is merged
+        indicesAll = []
+        for clad in self.parent.getChildrenWithFlags(flags.Flags.CLAD):
+            if isinstance(clad.spatialLocator, grids.MultiIndexLocation):
+                indicesAll.extend(clad.spatialLocator.indices)
+            else:
+                indicesAll.append(clad.spatialLocator.indices)
+
+        # Retrieve the indices of this component
+        indices = self.spatialLocator.indices
+        if not isinstance(indices, list):
+            indices = [indices]
+
+        # Map this component's indices to block's pin indices
+        getIndex = lambda ind: next(
+            (i for i, indA in enumerate(indicesAll) if np.all(ind == indA)),
+            None,
+        )
+        indexMap = list(map(getIndex, indices))
```

Comment:
This is kind of a lot. Walking through it, it seems to be doing what we want: generating a list of integers such that the components locator `self.spatialLocator.indices[i]` is found in the parent pin locations at
`self.parent.getPinLocations()[indexMap[i]]`

Do I have that right?

The inner iteration in the lambda func feels `O(n^2)`-like. I wonder if we could simplify with some considerations. I've got something like this cooked up in the assembly rotation where I need to make a mapping from pin locations to _something_. Here, we want to point each entry in `indices` to where it exists in `allIndices`, if at all.

I would recommend using a dictionary mapping the `(i, j, k)` tuples in `allIndices` to their integer location with something like 
```python
allLocations = self.parent.getPinLocations()
allIndices = {(loc.i, loc.j, loc.k): ix for ix, loc in enumerate(allLocations)}
indexMap = list(map(allIndices.get, indices))
```
There might be an intermediate step b/c `indices` isn't a tuple of ints but could easily be w/ (another!) map or something.

Spitballing, non-blocking comment.

Reply:
I'll defer to @zachmprince as he wrote this bit. I'll also pitch that we can forgo optimization for future work.